src/librustc_metadata/cstore_impl.rs

```diff
@@ -157,6 +243,98 @@ pub fn provide_local<'tcx>(providers: &mut Providers<'tcx>) {
 
     *providers = Providers {
         is_const_fn,
+        is_dllimport_foreign_item: |tcx, id| {
+            tcx.native_library_kind(id) == Some(NativeLibraryKind::NativeUnknown)
+        },
+        is_statically_included_foreign_item: |tcx, id| {
+            match tcx.native_library_kind(id) {
+                Some(NativeLibraryKind::NativeStatic) |
+                Some(NativeLibraryKind::NativeStaticNobundle) => true,
+                _ => false,
+            }
+        },
+        native_library_kind: |tcx, id| {
+            tcx.native_libraries(id.krate)
+                .iter()
+                .filter(|lib| native_libs::relevant_lib(&tcx.sess, lib))
+                .find(|l| l.foreign_items.contains(&id.index))
+                .map(|l| l.kind)
+        },
+        native_libraries: |tcx, cnum| {
+            assert_eq!(cnum, LOCAL_CRATE);
+            Rc::new(native_libs::collect(tcx))
+        },
+        link_args: |tcx, cnum| {
+            assert_eq!(cnum, LOCAL_CRATE);
+            Rc::new(link_args::collect(tcx))
+        },
+        extern_mod_stmt_cnum: |tcx, id| {
+            let id = tcx.hir.definitions().find_node_for_hir_id(id);
+            tcx.sess.cstore.extern_mod_stmt_cnum_untracked(id)
```

Comment:
This method and `postorder_cnums` below are something I wanted to ask about here. So what's happening here is that in this query it's reaching into the cstore via the `CrateStore` trait into various untracked methods to actually perform the computation.

Is that ok? Does that mean that these aren't tracked properly? Should the dep nodes be manually created here? Just wanted to make sure to flag this!

Reply:
cc https://github.com/rust-lang/rust/issues/44234 as well, where I believe we're discussing this basically